File: Chapter3.CoffeeShop/Models/Condiments/Soy.cs
namespace Chapter3.CoffeeShop.Models;

public class Soy : CondimentDecorator
{
    public Soy(Beverage beverage)
    {
        _beverage = beverage;
    }
    
    public override double Cost()
    {
        var price = 0.15;

        var cost = _beverage.GetSize() switch
        {
            Size.Tall => price - (price / 3),
            Size.Grande => price,
            Size.Venti => price + (price / 3),
            _ => throw new ArgumentOutOfRangeException()
        };

        return _beverage.Cost() + cost;
    }
    public override string GetDescription() => _beverage.GetDescription() + ", Soy";
}

File: Chapter3.CoffeeShop.Java/src/Condiments/Soy.java
package Condiments;

import Beverages.Beverage;

public class Soy extends CondimentDecorator{

    public Soy(Beverage beverage) {
        this.beverage = beverage;
    }

    @Override
    public double cost() {
        Double pricing = switch (beverage.getSize()) {
            case TALL -> 0.10;
            case GRANDE -> 0.15;
            case VENTI -> 0.20;
        };
        return beverage.cost() + pricing;
    }
    @Override
    public String getDescription() {
        return beverage.getDescription() + ", Soy";
    }
}
